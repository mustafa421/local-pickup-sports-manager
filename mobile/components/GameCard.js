import React from "react";
import PropTypes from "prop-types";
import {
  Text,
  StyleSheet,
  Alert,
  AsyncStorage,
  TouchableOpacity
} from "react-native";
import { connect } from "react-redux";
import { withNavigation } from "react-navigation";
import { Card, Button } from "react-native-elements";

const styles = StyleSheet.create({
  text: {
    marginBottom: 10
  }
});

const joinGame = async userInfo => {
  try {
    const request = await fetch(
      "http://local-pickup-sports-manager.herokuapp.com/joinGame",
      {
        method: "POST",
        headers: {
          "Content-Type": "application/json"
        },
        body: JSON.stringify(userInfo),
        json: true
      }
    );

    if (request.status !== 200) {
      throw Error("Failed to connect to join game");
    }

    let joinedGames = await AsyncStorage.getItem("joinedGames");

    if (!joinedGames) {
      const arr = [];
      await AsyncStorage.setItem("joinedGames", JSON.stringify(arr));
      joinedGames = arr;
    } else {
      joinedGames = JSON.parse(joinedGames);
    }

    if (!joinedGames.includes(userInfo.gameID)) {
      joinedGames.push(userInfo.gameID);
    }

    await AsyncStorage.setItem("joinedGames", JSON.stringify(joinedGames));

    Alert.alert("Success", "Successfully joined game", [{ text: "OK" }], {
      cancelable: false
    });
  } catch (ex) {
    console.log(ex);
    Alert.alert(
      "Unable to join game",
      "Error in attempting to join game. Please try again later.",
      [{ text: "OK" }],
      { cancelable: false }
    );
  }
};

const interestedGame = async userInfo => {
  try {
    const request = await fetch(
      "http://local-pickup-sports-manager.herokuapp.com/interestedGame",
      {
        method: "POST",
        headers: {
          "Content-Type": "application/json"
        },
        body: JSON.stringify(userInfo),
        json: true
      }
    );
    if (request.status !== 200) {
      throw new Error("Failed to send interested game");
    }

    let interestedGames = await AsyncStorage.getItem("interestedGames");

    if (!interestedGames) {
      const arr = [];
      await AsyncStorage.setItem("interestedGames", JSON.stringify(arr));
      interestedGames = arr;
    } else {
      interestedGames = JSON.parse(interestedGames);
    }

    // Maybe remove interest?
    if (!interestedGames.includes(userInfo.gameID)) {
      interestedGames.push(userInfo.gameID);
    }

    await AsyncStorage.setItem(
      "interestedGames",
      JSON.stringify(interestedGames)
    );

    Alert.alert(
      "Success",
      "Successfully showed interest in game.",
      [{ text: "OK" }],
      { cancelable: false }
    );
  } catch (err) {
    console.log(err);
    Alert.alert(
      "Unable to set interest",
      "Error in attempting to show interest in game. Please try again later.",
      [{ text: "OK" }],
      { cancelable: false }
    );
  }
};

export function GameCard(props) {
  const {
    title,
    skillLevel,
    duration,
    userID,
    username,
    sport,
    gameID,
    interested,
    joined,
    navigation,
    location,
    date
  } = props;

  return (
    <TouchableOpacity
      onPress={() =>
        navigation.navigate("gameScreen", {
          gameID,
          title,
          sport,
          location,
          date,
          skillLevel
        })
      }
    >
      <Card
        title={title}
        containerStyle={{
          backgroundColor: "#03A9F4",
          borderColor: "black"
        }}
        titleStyle={{
          fontWeight: "bold",
          fontSize: 20
        }}
      >
        <Text style={styles.text}>{sport} </Text>
        <Text style={styles.text}>{skillLevel}</Text>
        <Text style={styles.text}>Game Duration: {duration} </Text>

        <Button
          backgroundColor="paleturquoise"
          buttonStyle={{
            borderRadius: 5,
            marginLeft: 0,
            marginRight: 0,
            marginBottom: 0,
            borderColor: "black",
            borderWidth: 1
          }}
          title="Join Game"
          textStyle={{ color: "black" }}
          disabled={joined}
          onPress={() =>
            joinGame({
              userID,
              gameID,
              name: username
            })
          }
        />
        <Text style={{ marginBottom: 10 }} />

        <Button
          backgroundColor="paleturquoise"
          buttonStyle={{
            borderRadius: 5,
            marginLeft: 0,
            marginRight: 0,
            marginBottom: 0,
            borderColor: "black",
            borderWidth: 1
          }}
          title="Interested in Game"
          textStyle={{ color: "black" }}
          disabled={interested}
          onPress={() =>
            interestedGame({
              userID,
              gameID,
              name: username
            })
          }
        />
      </Card>
    </TouchableOpacity>
  );
}

const mapStateToProps = state => state.auth.userAccountData;

export default withNavigation(connect(mapStateToProps)(GameCard));

GameCard.propTypes = {
  title: PropTypes.string,
  sport: PropTypes.string,
  skillLevel: PropTypes.string.isRequired,
  duration: PropTypes.number.isRequired,
  userID: PropTypes.number.isRequired,
  gameID: PropTypes.number.isRequired,
  date: PropTypes.instanceOf(Date),
  location: PropTypes.string,
  interested: PropTypes.bool,
  joined: PropTypes.bool,
  username: PropTypes.string
};

GameCard.defaultProps = {
  username: null,
  sport: null,
  title: null,
  interested: false,
  joined: false,
  location: "",
  date: null
};
