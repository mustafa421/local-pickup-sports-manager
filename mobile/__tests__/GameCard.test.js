import React from "react";
import "react-native";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import MockAsyncStorage from "mock-async-storage";
import fetchMock from "fetch-mock";
import { AsyncStorage } from "react-native";
import { shallow } from "enzyme";
import Button from "./Button";
import renderer from "react-test-renderer";

import GameCard from "../components/GameCard";

// test join game (component is sending post requests)
describe("GameCard tests", () => {
  test("snapshot", () => {
    const tree = renderer.create(<GameCard />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
