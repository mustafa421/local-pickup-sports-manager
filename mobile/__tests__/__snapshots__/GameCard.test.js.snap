// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameCard tests snapshot 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "white",
          "borderColor": "#e1e8ee",
          "borderWidth": 1,
          "margin": 15,
          "marginBottom": 0,
          "padding": 15,
          "shadowColor": "rgba(0,0,0, .2)",
          "shadowOffset": Object {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 1,
          "shadowRadius": 1,
        },
        undefined,
        Object {
          "backgroundColor": "#03A9F4",
          "borderColor": "black",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
          },
          undefined,
          undefined,
        ]
      }
    >
      <View>
        <Text
          style={
            Array [
              Object {},
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              Array [
                Object {
                  "color": "#43484d",
                  "fontSize": 17.5,
                  "fontWeight": "bold",
                  "marginBottom": 15,
                  "textAlign": "center",
                },
                undefined,
                Object {
                  "fontSize": 20,
                  "fontWeight": "bold",
                },
                undefined,
              ],
            ]
          }
        >
          Basketball
        </Text>
        <View
          style={
            Array [
              Object {
                "backgroundColor": "#e1e8ee",
                "height": 1,
              },
              Array [
                Object {
                  "marginBottom": 15,
                },
                undefined,
              ],
            ]
          }
        />
      </View>
      <Text
        style={
          Object {
            "marginBottom": 10,
          }
        }
      >
         
      </Text>
      <Text
        style={
          Object {
            "marginBottom": 10,
          }
        }
      >
        Begginner
      </Text>
      <Text
        style={
          Object {
            "marginBottom": 10,
          }
        }
      >
        Game Duration: 
        1
         
      </Text>
      <View
        style={
          Array [
            Object {
              "marginLeft": 15,
              "marginRight": 15,
            },
            undefined,
            undefined,
            undefined,
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={null}
        >
          <View
            pointerEvents="box-only"
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#9E9E9E",
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "padding": 19,
                },
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                Object {
                  "backgroundColor": "paleturquoise",
                },
                undefined,
                Object {
                  "padding": 12,
                },
                undefined,
                undefined,
                undefined,
                Object {
                  "borderColor": "black",
                  "borderRadius": 5,
                  "borderWidth": 1,
                  "marginBottom": 0,
                  "marginLeft": 0,
                  "marginRight": 0,
                },
                false,
                false,
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {},
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  Array [
                    Object {
                      "color": "white",
                      "fontSize": 20,
                    },
                    undefined,
                    Object {
                      "fontSize": 17.5,
                    },
                    undefined,
                    undefined,
                    undefined,
                    undefined,
                    false,
                  ],
                ]
              }
            >
              Join Game
            </Text>
          </View>
        </View>
      </View>
      <Text
        style={
          Object {
            "marginBottom": 10,
          }
        }
      />
      <View
        style={
          Array [
            Object {
              "marginLeft": 15,
              "marginRight": 15,
            },
            undefined,
            undefined,
            undefined,
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={null}
        >
          <View
            pointerEvents="box-only"
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#9E9E9E",
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "padding": 19,
                },
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                Object {
                  "backgroundColor": "paleturquoise",
                },
                undefined,
                Object {
                  "padding": 12,
                },
                undefined,
                undefined,
                undefined,
                Object {
                  "borderColor": "black",
                  "borderRadius": 5,
                  "borderWidth": 1,
                  "marginBottom": 0,
                  "marginLeft": 0,
                  "marginRight": 0,
                },
                false,
                false,
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {},
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  Array [
                    Object {
                      "color": "white",
                      "fontSize": 20,
                    },
                    undefined,
                    Object {
                      "fontSize": 17.5,
                    },
                    undefined,
                    undefined,
                    undefined,
                    undefined,
                    false,
                  ],
                ]
              }
            >
              Interested in Game
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
